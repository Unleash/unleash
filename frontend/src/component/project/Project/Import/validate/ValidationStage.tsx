import { ImportLayoutContainer } from '../ImportLayoutContainer';
import { Box, Button, styled, Typography } from '@mui/material';
import React, { FC, useEffect, useState } from 'react';
import {
    IValidationSchema,
    useValidateImportApi,
} from 'hooks/api/actions/useValidateImportApi/useValidateImportApi';
import { ConditionallyRender } from 'component/common/ConditionallyRender/ConditionallyRender';
import useToast from 'hooks/useToast';
import { formatUnknownError } from 'utils/formatUnknownError';
import { ActionsContainer } from '../ActionsContainer';
import { IMPORT_CONFIGURATION_BUTTON } from 'utils/testIds';
import PermissionButton from 'component/common/PermissionButton/PermissionButton';
import { CREATE_FEATURE } from 'component/providers/AccessProvider/permissions';
import { usePlausibleTracker } from 'hooks/usePlausibleTracker';

const ImportInfoContainer = styled(Box)(({ theme }) => ({
    backgroundColor: theme.palette.background.elevation2,
    borderRadius: theme.shape.borderRadiusLarge,
    padding: theme.spacing(3),
}));

const Label = styled('span')(({ theme }) => ({
    fontSize: theme.fontSizes.smallBody,
    color: theme.palette.text.secondary,
}));
const Value = styled('span')(({ theme }) => ({
    fontSize: theme.fontSizes.smallBody,
    color: theme.palette.text.primary,
    fontWeight: theme.fontWeight.bold,
}));

const ErrorContainer = styled(Box)(({ theme }) => ({
    border: `1px solid ${theme.palette.error.border}`,
    borderRadius: theme.shape.borderRadiusLarge,
    paddingBottom: theme.spacing(2),
}));

const WarningContainer = styled(Box)(({ theme }) => ({
    border: `1px solid ${theme.palette.warning.border}`,
    borderRadius: theme.shape.borderRadiusLarge,
    paddingBottom: theme.spacing(2),
}));

const ErrorHeader = styled(Box)(({ theme }) => ({
    color: theme.palette.error.dark,
    backgroundColor: theme.palette.error.light,
    fontSize: theme.fontSizes.smallBody,
    borderBottom: `1px solid ${theme.palette.error.border}`,
    borderTopLeftRadius: theme.shape.borderRadiusLarge,
    borderTopRightRadius: theme.shape.borderRadiusLarge,
    padding: theme.spacing(2),
}));

const WarningHeader = styled(Box)(({ theme }) => ({
    color: theme.palette.warning.dark,
    backgroundColor: theme.palette.warning.light,
    fontSize: theme.fontSizes.smallBody,
    borderBottom: `1px solid ${theme.palette.warning.border}`,
    borderTopLeftRadius: theme.shape.borderRadiusLarge,
    borderTopRightRadius: theme.shape.borderRadiusLarge,
    padding: theme.spacing(2),
}));

const ErrorMessage = styled(Box)(({ theme }) => ({
    color: theme.palette.error.dark,
    fontSize: theme.fontSizes.smallBody,
}));

const WarningMessage = styled(Box)(({ theme }) => ({
    color: theme.palette.warning.dark,
    fontSize: theme.fontSizes.smallBody,
}));

const StyledItems = styled('ul')(({ theme }) => ({
    marginTop: theme.spacing(1),
    marginBottom: theme.spacing(0),
    paddingLeft: theme.spacing(3),
    paddingBottom: theme.spacing(3),
    borderBottom: `1px dashed ${theme.palette.neutral.border}`,
}));

const StyledItem = styled('li')(({ theme }) => ({
    fontSize: theme.fontSizes.smallBody,
}));

export const ValidationStage: FC<{
    environment: string;
    project: string;
    payload: string;
    onClose: () => void;
    onSubmit: () => void;
    onBack: () => void;
}> = ({ environment, project, payload, onClose, onBack, onSubmit }) => {
    const { validateImport } = useValidateImportApi();
    const { setToastData } = useToast();
    const { trackEvent } = usePlausibleTracker();
    const [validationResult, setValidationResult] = useState<IValidationSchema>(
        { errors: [], warnings: [], permissions: [] }
    );
    const [validJSON, setValidJSON] = useState(true);

    const trackValidation = (result: IValidationSchema) => {
        if (result.errors.length > 0 || result.permissions.length > 0) {
            trackEvent('export_import', {
                props: {
                    eventType: `validation fail`,
                },
            });
        } else {
            trackEvent('export_import', {
                props: {
                    eventType: `validation success`,
                },
            });
        }
        setValidationResult(result);
    };

    useEffect(() => {
        validateImport({ environment, project, data: JSON.parse(payload) })
            .then(trackValidation)
            .catch(error => {
                setValidJSON(false);
                setToastData({
                    type: 'error',
                    title: formatUnknownError(error),
                });
            });
    }, []);

    return (
        <ImportLayoutContainer>
            <ImportInfoContainer>
                <Typography sx={{ mb: 1.5 }}>
                    You are importing this configuration in:
                </Typography>
                <Box sx={{ display: 'flex', gap: 3 }}>
                    <span>
                        <Label>Environment: </Label>
                        <Value>{environment}</Value>
                    </span>
                    <span>
                        <Label>Project: </Label>
                        <Value>{project}</Value>
                    </span>
                </Box>
            </ImportInfoContainer>
            <ConditionallyRender
                condition={validationResult.permissions.length > 0}
                show={
                    <ErrorContainer>
                        <ErrorHeader>
                            <strong>Missing permissions!</strong> There are some
                            permissions that you need to be granted before
                            importing this configuration
                        </ErrorHeader>
                        {validationResult.permissions.map(error => (
                            <Box key={error.message} sx={{ p: 2 }}>
                                <ErrorMessage>{error.message}</ErrorMessage>
                                <StyledItems>
                                    {error.affectedItems.map(item => (
                                        <StyledItem key={item}>
                                            {item}
                                        </StyledItem>
                                    ))}
                                </StyledItems>
                            </Box>
                        ))}
                    </ErrorContainer>
                }
            />
            <ConditionallyRender
                condition={validationResult.errors.length > 0}
                show={
                    <ErrorContainer>
                        <ErrorHeader>
                            <strong>Conflict!</strong> There are some errors
                            that need to be fixed before the import.
                        </ErrorHeader>
                        {validationResult.errors.map(error => (
                            <Box key={error.message} sx={{ p: 2 }}>
                                <ErrorMessage>{error.message}</ErrorMessage>
                                <StyledItems>
                                    {error.affectedItems.map(item => (
                                        <StyledItem key={item}>
                                            {item}
                                        </StyledItem>
                                    ))}
                                </StyledItems>
                            </Box>
                        ))}
                    </ErrorContainer>
                }
            />
            <ConditionallyRender
                condition={validationResult.warnings.length > 0}
                show={
                    <WarningContainer>
                        <WarningHeader>
                            <strong>Warning!</strong> It is recommended to
                            verify the following information before importing.
                        </WarningHeader>
                        {validationResult.warnings.map(warning => (
                            <Box key={warning.message} sx={{ p: 2 }}>
                                <WarningMessage>
                                    {warning.message}
                                </WarningMessage>
                                <StyledItems>
                                    {warning.affectedItems.map(item => (
                                        <StyledItem key={item}>
                                            {item}
                                        </StyledItem>
                                    ))}
                                </StyledItems>
                            </Box>
                        ))}
                    </WarningContainer>
                }
            />
            <ActionsContainer>
                <Button
                    sx={{
                        position: 'static',
                        mr: 'auto',
                    }}
                    variant="outlined"
                    type="submit"
                    onClick={onBack}
                >
                    Back
                </Button>
                <PermissionButton
                    permission={CREATE_FEATURE}
                    projectId={project}
                    sx={{ position: 'static' }}
                    variant="contained"
                    type="submit"
                    onClick={onSubmit}
                    data-testid={IMPORT_CONFIGURATION_BUTTON}
                    disabled={
                        validationResult.errors.length > 0 ||
                        validationResult.permissions.length > 0 ||
                        !validJSON
                    }
                >
                    Import configuration
                </PermissionButton>
                <Button
                    sx={{ position: 'static', ml: 2 }}
                    variant="outlined"
                    type="submit"
                    onClick={onClose}
                >
                    Cancel import
                </Button>
            </ActionsContainer>
        </ImportLayoutContainer>
    );
};
