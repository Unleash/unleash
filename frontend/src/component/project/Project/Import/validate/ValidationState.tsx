import { ImportLayoutContainer } from '../ImportLayoutContainer';
import { Box, styled, Typography } from '@mui/material';
import { FC } from 'react';

const ImportInfoContainer = styled(Box)(({ theme }) => ({
    backgroundColor: theme.palette.secondaryContainer,
    borderRadius: theme.shape.borderRadiusLarge,
    padding: theme.spacing(3),
}));

const Label = styled('span')(({ theme }) => ({
    fontSize: theme.fontSizes.smallBody,
    color: theme.palette.text.secondary,
}));
const Value = styled('span')(({ theme }) => ({
    fontSize: theme.fontSizes.smallBody,
    color: theme.palette.text.primary,
    fontWeight: theme.fontWeight.bold,
}));

export const ValidationStage: FC<{ environment: string; project: string }> = ({
    environment,
    project,
}) => {
    return (
        <ImportLayoutContainer>
            <ImportInfoContainer>
                <Typography sx={{ mb: 1.5 }}>
                    You are importing this configuration in:
                </Typography>
                <Box sx={{ display: 'flex', gap: 3 }}>
                    <span>
                        <Label>Environment: </Label>
                        <Value>{environment}</Value>
                    </span>
                    <span>
                        <Label>Project: </Label>
                        <Value>{project}</Value>
                    </span>
                </Box>
            </ImportInfoContainer>
        </ImportLayoutContainer>
    );
};
