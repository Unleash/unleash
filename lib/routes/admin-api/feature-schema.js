'use strict';

const joi = require('joi');
const { nameType } = require('./util');

const nameSchema = joi.object().keys({ name: nameType });

const constraintSchema = joi.object().keys({
    contextName: joi.string(),
    operator: joi.string(),
    values: joi.array().items(joi.string()),
});

const strategiesSchema = joi.object().keys({
    name: nameType,
    constraints: joi
        .array()
        .allow(null)
        .items(constraintSchema),
    parameters: joi.object(),
});

const variantsSchema = joi.object().keys({
    name: nameType,
    weight: joi
        .number()
        .min(0)
        .max(1000)
        .required(),
    weightType: joi
        .string()
        .valid('variable', 'fix')
        .default('variable'),
    payload: joi
        .object()
        .keys({
            type: joi.string().required(),
            value: joi.string().required(),
        })
        .optional(),
    overrides: joi.array().items(
        joi
            .object()
            .keys({
                contextName: joi.string().required(),
                values: joi.array().items(joi.string()),
            })
            .optional(),
    ),
});

const featureShema = joi
    .object()
    .keys({
        name: nameType,
        enabled: joi.boolean().default(false),
        stale: joi.boolean().default(false),
        type: joi.string().default('release'),
        description: joi
            .string()
            .allow('')
            .allow(null)
            .optional(),
        strategies: joi
            .array()
            .required()
            .min(1)
            .items(strategiesSchema),
        variants: joi
            .array()
            .allow(null)
            .unique((a, b) => a.name === b.name)
            .optional()
            .items(variantsSchema),
    })
    .options({ allowUnknown: false, stripUnknown: true });

module.exports = { featureShema, strategiesSchema, nameSchema };
