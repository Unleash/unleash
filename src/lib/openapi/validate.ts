import Ajv, { ErrorObject } from 'ajv';
import { SchemaId, schemas } from './index';
import { omitKeys } from '../util/omit-keys';

interface ISchemaValidationErrors {
    schema: SchemaId;
    errors: ErrorObject[];
}

const ajv = new Ajv({
    schemas: Object.values(schemas).map((schema) =>
        omitKeys(schema, 'components'),
    ),

    // example was superseded by examples in openapi 3.1, but we're still on 3.0, so
    // let's add it back in!
    keywords: ['example'],
    formats: {
        'date-time': true,
        uri: true,
    },
});

export const validateSchema = (
    schema: SchemaId,
    data: unknown,
): ISchemaValidationErrors | undefined => {
    if (!ajv.validate(schema, data)) {
        return {
            schema,
            errors: ajv.errors ?? [],
        };
    }
};
