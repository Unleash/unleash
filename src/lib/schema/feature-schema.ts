import joi from 'joi';
import { ALL_OPERATORS } from '../util/constants';
import { nameType } from '../routes/util';
import { validateJsonString } from '../util/validateJsonString';

export const nameSchema = joi
    .object()
    .keys({ name: nameType })
    .options({ stripUnknown: true, allowUnknown: false, abortEarly: false });

export const constraintSchema = joi.object().keys({
    contextName: joi.string().required(),
    operator: joi
        .string()
        .valid(...ALL_OPERATORS)
        .required(),
    // Constraints must have a values array to support legacy SDKs.
    values: joi.array().items(joi.string().min(1).max(100)).default([]),
    value: joi.optional(),
    caseInsensitive: joi.boolean().optional(),
    inverted: joi.boolean().optional(),
});

export const strategiesSchema = joi.object().keys({
    id: joi.string().optional(),
    name: nameType,
    constraints: joi.array().allow(null).items(constraintSchema),
    parameters: joi.object(),
});

const variantValueSchema = joi
    .string()
    .required()
    // perform additional validation
    // when provided 'type' is 'json'
    .when('type', {
        is: 'json',
        then: joi.custom((val, helper) => {
            const isValidJsonString = validateJsonString(val);
            if (isValidJsonString === false) {
                return helper.error('invalidJsonString');
            }
            return val;
        }),
    })
    .messages({
        invalidJsonString:
            "'value' must be a valid json string when 'type' is json",
    });

export const variantsSchema = joi.object().keys({
    name: nameType,
    weight: joi.number().min(0).max(1000).required(),
    weightType: joi.string().valid('variable', 'fix').default('variable'),
    payload: joi
        .object()
        .keys({
            type: joi.string().required(),
            value: variantValueSchema,
        })
        .optional(),
    stickiness: joi.string().default('default'),
    overrides: joi.array().items(
        joi
            .object()
            .keys({
                contextName: joi.string().required(),
                values: joi.array().items(joi.string()),
            })
            .optional(),
    ),
});

export const variantsArraySchema = joi
    .array()
    .min(0)
    .items(variantsSchema)
    .unique((a, b) => a.name === b.name);

export const featureMetadataSchema = joi
    .object()
    .keys({
        name: nameType,
        stale: joi.boolean().default(false),
        archived: joi.boolean().default(false),
        type: joi.string().default('release'),
        description: joi.string().allow('').allow(null).optional(),
        impressionData: joi
            .boolean()
            .allow(true)
            .allow(false)
            .default(false)
            .optional(),
        createdAt: joi.date().optional().allow(null),
        variants: joi
            .array()
            .allow(null)
            .unique((a, b) => a.name === b.name)
            .optional()
            .items(variantsSchema),
    })
    .options({ allowUnknown: false, stripUnknown: true, abortEarly: false });

export const featureSchema = joi
    .object()
    .keys({
        name: nameType,
        enabled: joi.boolean().default(false),
        stale: joi.boolean().default(false),
        archived: joi.boolean().default(false),
        type: joi.string().default('release'),
        project: joi.string().default('default'),
        description: joi.string().allow('').allow(null).optional(),
        impressionData: joi
            .boolean()
            .allow(true)
            .allow(false)
            .default(false)
            .optional(),
        strategies: joi
            .array()
            .min(0)
            .allow(null)
            .optional()
            .items(strategiesSchema),
        variants: joi
            .array()
            .allow(null)
            .unique((a, b) => a.name === b.name)
            .optional()
            .items(variantsSchema),
    })
    .options({ allowUnknown: false, stripUnknown: true, abortEarly: false });

export const querySchema = joi
    .object()
    .keys({
        tag: joi
            .array()
            .allow(null)
            .items(joi.string().pattern(/\w+:.+/, { name: 'tag' }))
            .optional(),
        project: joi.array().allow(null).items(nameType).optional(),
        namePrefix: joi.string().allow(null).optional(),
        environment: joi.string().allow(null).optional(),
        inlineSegmentConstraints: joi.boolean().optional(),
    })
    .options({ allowUnknown: false, stripUnknown: true, abortEarly: false });

export const featureTagSchema = joi.object().keys({
    featureName: nameType,
    tagType: nameType.optional(),
    tagValue: joi.string(),
    type: nameType.optional(),
    value: joi.string().optional(),
});
