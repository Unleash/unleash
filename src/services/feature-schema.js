'use strict';

const joi = require('joi');
const { nameType } = require('../routes/admin-api/util');

const nameSchema = joi
    .object()
    .keys({ name: nameType })
    .options({ stripUnknown: true, allowUnknown: false, abortEarly: false });

const constraintSchema = joi.object().keys({
    contextName: joi.string(),
    operator: joi.string(),
    values: joi
        .array()
        .items(
            joi
                .string()
                .min(1)
                .max(100),
        )
        .min(1)
        .optional(),
});

const strategiesSchema = joi.object().keys({
    name: nameType,
    constraints: joi
        .array()
        .allow(null)
        .items(constraintSchema),
    parameters: joi.object(),
});

const variantsSchema = joi.object().keys({
    name: nameType,
    weight: joi
        .number()
        .min(0)
        .max(1000)
        .required(),
    weightType: joi
        .string()
        .valid('variable', 'fix')
        .default('variable'),
    payload: joi
        .object()
        .keys({
            type: joi.string().required(),
            value: joi.string().required(),
        })
        .optional(),
    overrides: joi.array().items(
        joi
            .object()
            .keys({
                contextName: joi.string().required(),
                values: joi.array().items(joi.string()),
            })
            .optional(),
    ),
});

const featureSchema = joi
    .object()
    .keys({
        name: nameType,
        enabled: joi.boolean().default(false),
        stale: joi.boolean().default(false),
        type: joi.string().default('release'),
        project: joi.string().default('default'),
        description: joi
            .string()
            .allow('')
            .allow(null)
            .optional(),
        strategies: joi
            .array()
            .required()
            .min(1)
            .items(strategiesSchema),
        variants: joi
            .array()
            .allow(null)
            .unique((a, b) => a.name === b.name)
            .optional()
            .items(variantsSchema),
    })
    .options({ allowUnknown: false, stripUnknown: true, abortEarly: false });

const querySchema = joi
    .object()
    .keys({
        tag: joi
            .array()
            .allow(null)
            .items(joi.string().pattern(/\w+:.+/, { name: 'tag' }))
            .optional(),
        project: joi
            .array()
            .allow(null)
            .items(joi.string().alphanum())
            .optional(),
        namePrefix: joi
            .string()
            .allow(null)
            .optional(),
    })
    .options({ allowUnknown: false, stripUnknown: true, abortEarly: false });

module.exports = { featureSchema, strategiesSchema, nameSchema, querySchema };
