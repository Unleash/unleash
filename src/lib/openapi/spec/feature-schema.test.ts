import { validateSchema } from '../validate';
import { FeatureSchema } from './feature-schema';

test('featureSchema', () => {
    const data: FeatureSchema = {
        name: 'a',
        strategies: [
            {
                id: 'a',
                name: 'a',
                constraints: [
                    {
                        contextName: 'a',
                        operator: 'IN',
                    },
                ],
                segments: [1],
            },
        ],
        variants: [
            {
                name: 'a',
                weight: 1,
                weightType: 'a',
                stickiness: 'a',
                overrides: [{ contextName: 'a', values: ['a'] }],
                payload: { type: 'a', value: 'b' },
            },
        ],
        environments: [
            {
                name: 'a',
                type: 'b',
                enabled: true,
            },
        ],
    };

    expect(
        validateSchema('#/components/schemas/featureSchema', data),
    ).toBeUndefined();
});

test('featureSchema constraints', () => {
    const data = {
        name: 'a',
        strategies: [{ name: 'a', constraints: [{ contextName: 'a' }] }],
    };

    expect(
        validateSchema('#/components/schemas/featureSchema', data),
    ).toMatchSnapshot();
});

test('featureSchema variants should only have a few required fields', () => {
    const data = {
        name: 'a',
        variants: [
            {
                name: 'a',
                weight: 1,
            },
        ],
    };

    expect(
        validateSchema('#/components/schemas/featureSchema', data),
    ).toBeUndefined();
});

test('featureSchema variant override values must be an array', () => {
    const data = {
        name: 'a',
        variants: [
            {
                name: 'a',
                weight: 1,
                weightType: 'a',
                stickiness: 'a',
                overrides: [{ contextName: 'a', values: 'b' }],
                payload: { type: 'a', value: 'b' },
            },
        ],
    };

    expect(
        validateSchema('#/components/schemas/featureSchema', data),
    ).toMatchSnapshot();
});
