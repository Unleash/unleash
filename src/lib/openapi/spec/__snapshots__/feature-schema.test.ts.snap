// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`featureSchema constraints 1`] = `
{
  "errors": [
    {
      "instancePath": "/strategies/0/constraints/0",
      "keyword": "required",
      "message": "must have required property 'operator'",
      "params": {
        "missingProperty": "operator",
      },
      "schemaPath": "#/components/schemas/constraintSchema/required",
    },
  ],
  "schema": "#/components/schemas/featureSchema",
}
`;

exports[`featureSchema variant override values must be an array 1`] = `
{
  "errors": [
    {
      "instancePath": "/variants/0/overrides/0/values",
      "keyword": "type",
      "message": "must be array",
      "params": {
        "type": "array",
      },
      "schemaPath": "#/components/schemas/overrideSchema/properties/values/type",
    },
  ],
  "schema": "#/components/schemas/featureSchema",
}
`;
